### PR TITLE
Periodically clear JarResource when not accessed

### PR DESCRIPTION
**Pull request by geoand**
Relates to: #13258

### CODE REVIEW

**Comment by geoand** [comment]
@Sanne this is a draft implementation of the timer based idea discussed in #13258

File: independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/SerializedApplication.java [hunk 1]
```diff
@@ -115,12 +115,26 @@ public static SerializedApplication read(InputStream inputStream, Path appRoot)
...
                 parentFirstPackages.add(in.readUTF());
             }
+
             return new SerializedApplication(
-                    new RunnerClassLoader(ClassLoader.getSystemClassLoader(), resourceDirectoryMap, parentFirstPackages),
+                    new RunnerClassLoader(ClassLoader.getSystemClassLoader(), resourceDirectoryMap, parentFirstPackages,
+                            new RunnerClassLoader.CleanerConfig(
+                                    !"true".equals(System.getProperty("quarkus.runner.cleaner.prevent")),
+                                    getCleanerSystemProperty("quarkus.runner.cleaner.period", 30_000),
+                                    getCleanerSystemProperty("quarkus.runner.cleaner.unaccessed", 20_000))),
```
Comment (geoand) [inline independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/SerializedApplication.java:124]: These defaults as well as the names are obviously lame...

**Comment by gsmet** [comment]
My first inclination would have been to do it only once after the bootstrap (well, let's say 2 minutes after the end of the bootstrap for instance), so that we don't clean up resources when they are accessed post-startup.

What I'm a bit afraid of is a scenario when we close resources/open resources regularly. I'm not sure this will be a win?

Or maybe several times during the first 10 minutes of execution but then I would have let the app live its life.

Not sure if it makes sense?

**Comment by geoand** [comment]
> My first inclination would have been to do it only once after the bootstrap (well, let's say 2 minutes after the end of the bootstrap for instance), so that we don't clean up resources when they are accessed post-startup.
> 

We do do this in some form, but it takes place a little early.

> What I'm a bit afraid of is a scenario when we close resources/open resources regularly. I'm not sure this will be a win?
> 

It shouldn't happen if we do like this PR does, where only non-accessed jars are closed. I believe if we come up with the right defaults, we should cover most real world cases.

> Or maybe several times during the first 10 minutes of execution but then I would have let the app live its life.
> 

I am not sure what's best TBH. I am not sure what the proper defaults should be, but I do think that making this periodic makes sense.
Of course this could also be configurable.

> Not sure if it makes sense?

It sure does. I just don't know what the proper defaults should be :)

**Comment by Sanne** [comment]
> My first inclination would have been to do it only once after the bootstrap (well, let's say 2 minutes after the end of the bootstrap for instance), so that we don't clean up resources when they are accessed post-startup.

Imagine this scenario: you have a job scheduled to trigger after 10 minutes; when it's first triggered it will need to load its classes, and it will possibly do something silly such as list all resources in `META-INF`, so to initialize something.. this will get us to re-open all jars and never close them again.
Ideally, we're able to recover from this, trigger the close, and at the next run supposedly that code was already loaded.

**Comment by geoand** [comment]
@stuartwdouglas WDYT about this?

**Comment by Sanne** [comment]
Would you mind rebasing @geoand ?

**Comment by geoand** [comment]
@Sanne done

**Comment by geoand** [comment]
And again to pick up the latest merged changed as well

**Comment by Sanne** [comment]
I'm not really liking this approach much, sorry :/

We can't expect people to go tuning such aspects, for such a very small optimisation. We need a better idea with ergonomics in mind, or at least some approach for which we could feel comofortable in hard-coding a reasonable strategy.

Also I don't want to start a thread just for this purpose - it's overkill.

What if we were to do some cleanup on a new resource being loaded? Most resource loading will happen in bursts of sequential access to the same Jar - especially now that you've optimised the resource indexing. A little "dancing" between dependencies should be tolerated, but essentially we can opt to keep only N jar files open, with N being a reasonably low number (3? 2?).

When load activities cease, we'll be stuck with N jars whose head is buffered - no big deal, and still far better than 100+.

**Comment by stuartwdouglas** [comment]
I like the approach proposed by @Sanne better, I don't really like the idea of starting a timer.

**Comment by Sanne** [comment]
here is a quick sketch to illustrate what I have in mind - totally untested yet:

 - https://github.com/quarkusio/quarkus/commit/e044b795378d7ba4e11cedd7b7c5bc56f45dd03d

**Comment by stuartwdouglas** [comment]
Here is an alternative idea, if any resources are accessed multiple times after boot we could unpack them to a temp dir. This means it would not be required to hold onto the jar file, and could also improve HTTP perf if they are web resources as you could potentially do zero copy transfer.

**Comment by Sanne** [comment]
Sure but that's for resources.. we should do it.  But in this case we're looking at class files which have been loaded later.

**Comment by Sanne** [comment]
Actually had some fun implementing it:
 - https://github.com/quarkusio/quarkus/commit/13f3352e00d5be5b6c47008198780478ec05b0cb

memory profiles look very good. Didn't run tests yet though, and should measure what's the penalty on bootstrap times.

**Comment by geoand** [comment]
So I'll go ahead and close this.

I agree it's pretty crude and the ideas you guys brought up are better

**Pull request closed without merging**